src/Featurizers/TruncatedSVDFeaturizer.h

```diff
@@ -0,0 +1,631 @@
+};
+
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// |
+// |  Implementation
+// |
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+
+// ----------------------------------------------------------------------
+// |
+// |  SVDComponentsAnnotationData
+// |
+// ----------------------------------------------------------------------
+template <typename T>
+SVDComponentsAnnotationData<T>::SVDComponentsAnnotationData(T singularvalues, T singularvectors) :
+    SingularValues(
+        std::move(
+            [&singularvalues](void) -> T & {
+                if(singularvalues.size() == 0)
+                    throw std::invalid_argument("singularvalues");
+
+                return singularvalues;
+            }()
+        )
+    ),
+    SingularVectors(
+        std::move(
+            [&singularvectors](void) -> T & {
+                if(singularvectors.size() == 0)
+                    throw std::invalid_argument("singularvectors");
+
+                return singularvectors;
+            }()
+        )
+    ) {
+}
+
+// ----------------------------------------------------------------------
+// |
+// |  SVDComponentsEstimator
+// |
+// ----------------------------------------------------------------------
+template <typename InputT, typename TransformedT, size_t MaxNumTrainingItemsV>
+SVDComponentsEstimator<InputT, TransformedT, MaxNumTrainingItemsV>::SVDComponentsEstimator(AnnotationMapsPtr pAllColumnAnnotations, size_t colIndex) :
+    BaseType(std::move(pAllColumnAnnotations), std::move(colIndex), true) {
+}
+
+template <typename InputT, typename TransformedT>
+void ComponentsDetails::SVDTrainingOnlyPolicy<InputT, TransformedT>::fit(InputType const &input) {
+
+    //assume same type but remain InputT and TransformedT for flexibility
+    if (!std::is_same<InputT,TransformedT>::value)
+        throw std::invalid_argument("InputT should be equal with TransformedT");
+    
+    _matrix = input;
+}
+
+template <typename InputT, typename TransformedT>
+SVDComponentsAnnotationData<TransformedT> ComponentsDetails::SVDTrainingOnlyPolicy<InputT, TransformedT>::complete_training(void) {
+
+    //the following code in this function is introduced from RedSVD
+    typedef typename TransformedT::Scalar Scalar;
+    typedef typename Eigen::Matrix<Scalar, Eigen::Dynamic, Eigen::Dynamic> DenseMatrix;
+
+    Eigen::Index rank = (_matrix.rows() < _matrix.cols()) ? _matrix.rows() : _matrix.cols();
+
+    // Gaussian Random Matrix for _matrix^T
+	DenseMatrix O(_matrix.rows(), rank);
+	sample_gaussian(O);
+			
+	// Compute Sample Matrix of _matrix^T
+	DenseMatrix Y = _matrix.transpose() * std::move(O);
```

Comment:
>std::move(O) [](start = 39, length = 12)

What's the purpose of the move? Same below.

Reply:
Just save some spaces as the variable will no longer needed

Reply:
This part I do not understand.

---
In reply to: [366111577](https://github.com/microsoft/FeaturizersLibrary/pull/64#discussion_r366111577) [](ancestors = 366111577)